applications/FluidDynamicsApplication/custom_conditions/adjoint_monolithic_wall_condition.cpp

```diff
@@ -0,0 +1,779 @@
+    std::size_t Step)
+{
+    auto& r_node = mpCondition->GetGeometry()[NodeId];
+    rVector.resize(4);
+    std::size_t index = 0;
+    rVector[index++] = MakeIndirectScalar(r_node, ADJOINT_FLUID_VECTOR_3_X, Step);
+    rVector[index++] = MakeIndirectScalar(r_node, ADJOINT_FLUID_VECTOR_3_Y, Step);
+    rVector[index++] = MakeIndirectScalar(r_node, ADJOINT_FLUID_VECTOR_3_Z, Step);
+    rVector[index] = IndirectScalar<double>{}; // pressure
+}
+
+template <>
+void AdjointMonolithicWallCondition<2, 2>::ThisExtensions::GetAuxiliaryVector(
+    std::size_t NodeId,
+    std::vector<IndirectScalar<double>>& rVector,
+    std::size_t Step)
+{
+    auto& r_node = mpCondition->GetGeometry()[NodeId];
+    rVector.resize(3);
+    std::size_t index = 0;
+    rVector[index++] = MakeIndirectScalar(r_node, AUX_ADJOINT_FLUID_VECTOR_1_X, Step);
+    rVector[index++] = MakeIndirectScalar(r_node, AUX_ADJOINT_FLUID_VECTOR_1_Y, Step);
+    rVector[index] = IndirectScalar<double>{}; // pressure
+}
+
+template <>
+void AdjointMonolithicWallCondition<3, 3>::ThisExtensions::GetAuxiliaryVector(
+    std::size_t NodeId,
+    std::vector<IndirectScalar<double>>& rVector,
+    std::size_t Step)
+{
+    auto& r_node = mpCondition->GetGeometry()[NodeId];
+    rVector.resize(4);
+    std::size_t index = 0;
+    rVector[index++] = MakeIndirectScalar(r_node, AUX_ADJOINT_FLUID_VECTOR_1_X, Step);
+    rVector[index++] = MakeIndirectScalar(r_node, AUX_ADJOINT_FLUID_VECTOR_1_Y, Step);
+    rVector[index++] = MakeIndirectScalar(r_node, AUX_ADJOINT_FLUID_VECTOR_1_Z, Step);
+    rVector[index] = IndirectScalar<double>{}; // pressure
+}
+
+template <unsigned int TDim, unsigned int TNumNodes>
+void AdjointMonolithicWallCondition<TDim, TNumNodes>::ThisExtensions::GetFirstDerivativesVariables(
+    std::vector<VariableData const*>& rVariables) const
+{
+    rVariables.resize(1);
+    rVariables[0] = &ADJOINT_FLUID_VECTOR_2;
+}
+
+template <unsigned int TDim, unsigned int TNumNodes>
+void AdjointMonolithicWallCondition<TDim, TNumNodes>::ThisExtensions::GetSecondDerivativesVariables(
+    std::vector<VariableData const*>& rVariables) const
+{
+    rVariables.resize(1);
+    rVariables[0] = &ADJOINT_FLUID_VECTOR_3;
+}
+
+template <unsigned int TDim, unsigned int TNumNodes>
+void AdjointMonolithicWallCondition<TDim, TNumNodes>::ThisExtensions::GetAuxiliaryVariables(
+    std::vector<VariableData const*>& rVariables) const
+{
+    rVariables.resize(1);
+    rVariables[0] = &AUX_ADJOINT_FLUID_VECTOR_1;
+}
+
+template <>
+void AdjointMonolithicWallCondition<2, 2>::EquationIdVector(
+    EquationIdVectorType& rResult,
+    const ProcessInfo& rCurrentProcessInfo) const
+{
+    if (rResult.size() != 6) {
+        rResult.resize(6);
+    }
+
+    IndexType local_index = 0;
+    for (IndexType i = 0; i < 2; ++i) {
+        const auto& r_node = this->GetGeometry()[i];
+        rResult[local_index++] = r_node.GetDof(ADJOINT_FLUID_VECTOR_1_X).EquationId();
```

Comment:
Use the `GetDof` with position. It is way faster.

Reply:
(this comment applies to the other methods)

Reply:
done :)